Bump Serialized DAG to v2 and handle conversion from v1

dstandish opened the pull request:
A number of changes have been made to the serialized DAG structure between 2.x and 3.0, and with the addition of DAG versioning in 3.0 we can't use our old approach of simply force-reserializing all DAGs anymore.

So instead we bump the version record (that has been `1` since it was first introduced in 1.10.7, even though there were changes) and upgrade from the latest format in Airflow 2.x such that the Webserver will be able to display things for it.

The scheduler is not so easy to fix, so we use a proxy of "skip things that haven't been re-processed" yet by making the scheduler ignore DAGs that don't a bundle name set (as all dags processed in 3.0 have this set, even if it is for a local file bundle)

dstandish committed:
Handle v1 serdag conversion

dstandish committed:
get it so scheduler and webserver don't immediately crash

dstandish committed:
some fixes

ephraimbuddy committed:
Fix serialization test as we have set version to 2

ephraimbuddy commented on airflow-core/src/airflow/serialization/serialized_objects.py:
```diff
@@ -1690,7 +1691,7 @@ def serialize_dag(cls, dag: DAG) -> dict:
             raise
         except Exception as e:
-            raise SerializationError(f"Failed to serialize DAG {dag.dag_id!r}: {e}")
+            raise SerializationError(f"Failed to serialize DAG {dag.dag_id!r}") from e
```
Returning this to the original because it has more info. Maybe we can change this in another PR

ephraimbuddy committed:
Make deserialized dataset dags compatible with Airflow 3

ashb commented on airflow-core/src/airflow/jobs/scheduler_job_runner.py:
```diff
@@ -356,6 +356,7 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session) -
                 .join(TI.dag_model)
                 .where(~DM.is_paused)
                 .where(TI.state == TaskInstanceState.SCHEDULED)
+                .where(DagModel.bundle_name.is_not(None))
```
Let be consistent with elsewhere in the query
```suggestion
                .where(DM.bundle_name.is_not(None))
```

This is so the scheduler won't look at dags that haven't been re-serialized yet for AF3, yeah?

ashb commented on airflow-core/src/airflow/serialization/serialized_objects.py:
```diff
@@ -1135,7 +1135,8 @@ def detect_dag_dependencies(dag: DAG | None) -> Iterable[DagDependency]:
             return
-
+        if not dag.timetable:
+            return
```
You shouldn't really end up with no timetable -- shouldn't you get a `NullTimetable` object instead?

ashb replied:
```suggestion
```

Should be handled now

ashb commented on airflow-core/src/airflow/api_fastapi/core_api/services/ui/grid.py:
```diff
@@ -275,6 +278,13 @@ def _get_serdag(ti, session):
+            dag_id=dag_version.dag_id,
+            version_id=dag_version.id,
+            version_number=dag_version.version_number,
+        )
```
We log an error but return None. Is that okay?

kaxil replied:
maybe warn instead of error?

ashb committed:
Simplify DAG conversion code

- Make it less general purpose. YAGNI (You Ain't Gonna Need it) -- we can add
  it back in once we know what patterns make sense

- Make each conversion function deal with exactly one version, from it's
  version to the next. That way as the number of versions grow, we don't need
  to handle all new versions in all old migrations, we can just chaing
  v1_to_v2 -> v2_to_v3 etc.

- Upgrade schedule_interval to a timetable -- it is not complete yet

ashb committed:
Upgrade more old schedule interval types

ashb committed:
Update airflow-core/src/airflow/serialization/serialized_objects.py

ashb committed:
Update airflow-core/src/airflow/jobs/scheduler_job_runner.py

ashb committed:
Upgrade old DatasetOrTime timetable

ashb committed:
Update `@continuous` too

ashb committed:
Update more cases

ashb committed:
Add tests for serdag upgrade

kaxil commented on airflow-core/src/airflow/serialization/serialized_objects.py:
```diff
@@ -1787,12 +1788,140 @@ def to_dict(cls, var: Any) -> dict:
+            ("_access_control", "access_control"),
+        ]
+        task_renames = [("_task_type", "task_type")]
+        #
```
```suggestion
```

kaxil commented on airflow-core/src/airflow/serialization/serialized_objects.py:
```diff
@@ -1787,12 +1788,140 @@ def to_dict(cls, var: Any) -> dict:
+                    # We should maybe convert this to None and warn instead
+                    raise ValueError(f"Unknown schedule_interval field {sched!r}")
+            elif sched.get("__type") == "timedelta":
+                dag_dict["timetable"] = {
+                    "__type": "airflow.timetables.interval.DeltaDataIntervalTimetable",
+                    "__var": {"delta": sched["__var"]},
+                }
+        elif timetable := dag_dict.get("timetable"):
+            if timetable["__type"] in {
+                "airflow.timetables.simple.DatasetTriggeredTimetable",
+                "airflow.timetables.datasets.DatasetOrTimeSchedule",
+            }:
+                dag_dict["timetable"] = _replace_dataset_with_asset_in_timetables(dag_dict["timetable"])
+
+        if "dag_dependencies" in dag_dict:
+            for dep in dag_dict["dag_dependencies"]:
+                for fld in ("dependency_type", "target", "source"):
+                    if dep.get(fld) == "dataset":
+                        dep[fld] = "asset"
+
+        for task in dag_dict["tasks"]:
+            task_var: dict = task["__var"]
+            if "airflow.ti_deps.deps.ready_to_reschedule.ReadyToRescheduleDep" in task_var.get("deps", []):
+                task_var["_is_sensor"] = True
+            for k in tasks_remove:
+                task_var.pop(k, None)
+            for old, new in task_renames:
+                task_var[new] = task_var.pop(old)
+            for item in task_var.get("outlets", []):
+                if isinstance(item, dict) and "__type" in item:
+                    item["__type"] = replace_dataset_in_str(item["__type"])
+            for item in task_var.get("inlets", []):
+                if isinstance(item, dict) and "__type" in item:
+                    item["__type"] = replace_dataset_in_str(item["__type"])
+                var_ = item["__var"]
+                var_["name"] = None
+                var_["group"] = None
+
+        # Set on the root TG
+        dag_dict["task_group"]["group_display_name"] = ""
```
Probably should separate it out into a different file and have a better way of handling it later on -- but fine for now
